stor/dx.py

```diff
@@ -113,96 +165,727 @@ def temp_url(self, lifetime=300, filename=None):
+            DNAnexusError: When copying within same project with raise_if_same_project=False
+            NotFoundError: When the source file path doesn't exist
+        """
+        dest = Path(dest)
+        if utils.is_dx_path(dest):
+            if self.isfile():
+                if dest.canonical_project == self.canonical_project:
+                    if not raise_if_same_project:
+                        self._move(dest)
+                    else:
+                        raise DNAnexusError('Source and destination are in same project. '
+                                            'Set raise_if_same_project=False to allow this.')
+                else:
+                    self._clone(dest)
+            else:
+                raise stor_exceptions.NotFoundError(
+                    'No data object was found for the given path on DNAnexus')
         else:
-            return self.stat()['size']
+            super(DXPath, self).copy(dest)  # for other filesystems, delegate to utils.copy
+
+    def copytree(self, dest, raise_if_same_project=False, **kwargs):
+        """Copies a source directory to a destination directory.
+        This is not an atomic operation.
 
-    def download_object(self, dest):
-        """Download a single path or object to file."""
-        raise NotImplementedError
+        If the destination path already exists as a directory, the source tree
+        including the root folder is copied over as a subfolder of the destination.
 
-    def download(self, dest):
-        """Download a directory."""
-        raise NotImplementedError
+        If the source and destination directories belong to the same project,
+        the tree is moved instead of copied. Also, in such cases, the root folder
+        of the project cannot be the source path. Please listdir the root folder
+        and copy/copytree individual items if needed.
 
-    def upload(self, source):
-        """Upload a list of files and directories to a directory."""
-        raise NotImplementedError
+        For example, assume the following file hierarchy::
 
-    def open(self, mode='r', encoding=None):
+            project1/
+            - b/
+            - - 1.txt
+
+            project2/
+
+        Doing a copytree from ``project1:/b/`` to a new dx destination of
+        ``project2:/c`` is performed with::
+
+            Path('dx://project1:/b').copytree('dx://project2:/c')
+
+        The end result for project2 looks like::
+
+            project2/
+            - c/
+            - - 1.txt
+
+        If the destination path directory already exists, the folder is copied
+        as a subfolder of the destination. If this new destination also exists,
+        a TargetExistsError is raised.
+
+        If the source ris a root folder, and is cloned to an existing destination directory
+        or if the destination is also a root folder, the tree is moved under project name.
+
+        Refer to utils.copytree for detailed information.
```

Comment:
is there actually more detailed information there? seems like it's actually in DX docs and you should cross-reference

Reply:
Okay.